### PR TITLE
Detect untyped top level functions.

### PR DESCRIPTION
**Pull request by nojaf**
WIP, this tries to detect if a top-level function is fully typed or not.

### CODE REVIEW

**Comment by nojaf** [comment]
@dawedawe, I can process `Fantomas.Core` with the current state of this analyzer.
Could you please skim through the code and point out where the readability can improve in the code can improve?

But hold back, please point out where things are less clear.

File: src/FSharp.Analyzers/TopLevelTypeAnalysis.fs [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+        | SynTyparDecls.PostfixList (decls = decls ; constraints = constraints) ->
+            let initialMap =
+                decls
+                |> List.map (fun (SynTyparDecl (typar = SynTypar (ident = ident))) ->
+                    ident.idText, UntypedGenericParameterInfo.Empty
+                )
+                |> Map.ofList
+
+            (initialMap, constraints)
+            ||> List.fold (fun map stc ->
```
Comment (dawedawe) [inline src/FSharp.Analyzers/TopLevelTypeAnalysis.fs:166]: Maybe a less generic name for the map would be nice.

File: src/FSharp.Analyzers/TopLevelTypeAnalysis.fs [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+                match stc with
+                | SynTypeConstraint.WhereTyparIsValueType (typar, range) ->
+                    failwithf "todo: SynTypeConstraint.WhereTyparIsValueType %A %s" range range.FileName
```
Comment (dawedawe) [inline src/FSharp.Analyzers/TopLevelTypeAnalysis.fs:169]: I think, later on, when all the todos are implemented, a bit of refactoring to keep the match construct to a reasonable amount of lines would be good.

**Comment by dawedawe** [comment]
> @dawedawe, I can process `Fantomas.Core` with the current state of this analyzer. Could you please skim through the code and point out where the readability can improve in the code can improve?
> 
> But hold back, please point out where things are less clear.

Impressive work.
So far, there's no refactoring needed imho.

**Comment by nojaf** [comment]
No longer pursuing this.

**Pull request closed without merging**